File: ios/Earwyrm/Core/Networking/LRCLIBService.swift
import Foundation

/// Fetch full lyrics from LRCLIB (free, no auth required).
enum LRCLIBService {
    private static let baseURL = "https://lrclib.net/api/get"

    /// Fetch plain lyrics for a given artist and track title.
    /// Returns `nil` on any error (404, network, decode).
    static func fetchLyrics(artist: String, title: String) async -> String? {
        let trimmedArtist = artist.trimmingCharacters(in: .whitespaces)
        let trimmedTitle = title.trimmingCharacters(in: .whitespaces)
        guard !trimmedArtist.isEmpty, !trimmedTitle.isEmpty else { return nil }

        var components = URLComponents(string: baseURL)
        components?.queryItems = [
            URLQueryItem(name: "artist_name", value: trimmedArtist),
            URLQueryItem(name: "track_name", value: trimmedTitle),
        ]

        guard let urlString = components?.url?.absoluteString else { return nil }

        do {
            let response: LRCLIBResponse = try await HTTPClient.get(urlString)
            return response.plainLyrics
        } catch {
            return nil
        }
    }
}

private struct LRCLIBResponse: Decodable {
    let plainLyrics: String?
    let syncedLyrics: String?
}

File: android/app/src/main/java/com/earwyrm/app/core/network/LRCLIBService.kt
package com.earwyrm.app.core.network

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.get
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class LRCLIBService @Inject constructor(private val httpClient: HttpClient) {
    suspend fun getLyrics(title: String, artist: String): String? {
        return try {
            val response = httpClient.get("https://lrclib.net/api/get") {
                url { parameters.append("track_name", title); parameters.append("artist_name", artist) }
            }
            val result = response.body<LRCLIBResponse>()
            result.plainLyrics ?: result.syncedLyrics
        } catch (_: Exception) { null }
    }

    @Serializable private data class LRCLIBResponse(@SerialName("plainLyrics") val plainLyrics: String? = null, @SerialName("syncedLyrics") val syncedLyrics: String? = null)
}
